### ISSUE TITLE
Calculating post-receptoral contrasts

### ISSUE THREAD

**Issue by spitschan** [spitschan/SilentSubstitutionToolbox#19]
We've had a few discussions about the best way to calculate post-receptoral contrasts around arbitrary backgrounds. Here's what we've done in the past:

> We transformed LMS cone contrasts to a post-receptoral opponent representation assuming mechanism sensitivities to cone contrast for luminance, red-green, and blue-yellow mechanisms of [0.5 0.5 0], [0.5 -0.5 0], and [-0.5 -0.5 1] respectively.  This transformation corresponds to the DKL opponent color space representation (ref DKL paper, my chapter) when the background produces equal excitations in the L, M and S cones, for the case where the L and M cone spectral sensitivities are scaled so that they sum to produce the luminous efficiency curve. We regarded this as a a reasonable choice of reference conditions to define the transformation, as it leads to intuitively straightforward mechanism properties. For other backgrounds, this transformation will describe the opponent mechanism responses to the extent that those responses are the same for modulations seen against different backgrounds when their LMS cone contrasts are matched.

Further from DHB:

> This choice has the advantage that it seems about as reasonable as anything else, matches what we have been doing and the the pre-reg document, and is explainable (as above) even if the explanation hides many subtleties.  I think the property that we expect iso-response contours to better line up with the axes of an opponent space justifies the extra complication for our present purposes.

Opening this as an issue to see if we might want to write a function for this, or address the transformation programmatically in some way.

**Comment by spitschan** [comment]
Also related to https://github.com/spitschan/SilentSubstitutionToolbox/issues/10

**Comment by spitschan** [comment]
Bumping this for visibility

**Comment by DavidBrainard** [comment]
I wrote ComputePostreceptoralContrastsFromLMSContrasts which handles the post-receptoral contrast calcs as we did in the MaxMel paper.  Still need to work on a demo, integrating with the SST splatter plots and lightcalcs, and then into the OL stuff.  So leaving this issue open.

**Comment by spitschan** [comment]
Perfect.

I recently stripped code from `OLCheckLightLevels` and moved it into a routine in `SilentSubstitutionToolbox/SST/lightcalcs/CalculateLightProperties.m` which takes the spectral radiance measurements from the PR-670 instruments and calculates the relevant quantities. Would be worth integrating this as well. Let me know if I can help you integrate this.

**Comment by spitschan** [comment]
Is this solved @DavidBrainard?

**Comment by spitschan** [comment]
Did we ever find an intellectual resolution to this, @DavidBrainard and @JorisVincent?

**Comment by JorisVincent** [comment]
I don't think we did. My thought would be to separate some of the functionality, i.e.:

1. Have one function that `computes` the postreceptoral contrasts (`ComputePostreceptoralContrasts`). I envision such a function to take in a set of receptor contrasts, and some matrix defining the postreceptoral combinations. This function would be trivial (one matrix multiplication), but could do some input verification to make sure you have the orientation correct etc.
2. Have one function that `reports` the postreceptoral contrasts (`ReportReceptoralContrasts`), which takes the output from (1), and string array with the labels for the contrasts, and prints this. This function could likely be the same for postreceptoral and receptoral contrasts.
3. Have one or more functions that generate standard transformation matrices for (1) (`GetPostreceptoralCombinations`). I envision this function to take in a string, such as `DKL`, and return a matrix defining how to combine the receptor contrasts, which could then be plugged into (1).

Combined, you could replicate the current behavior which something like:
`ReportReceptoralContrasts( ComputePostreceptoralContrasts(LMSContrasts, GetPostreceptoralCombinations('DKL') ) , {'L+M+S', 'L-M', 'S-(L+M)'} );`
but this would also allow us to do different postreceptoral combinations, labels, etc.

**Comment by spitschan** [comment]
This seems like a sensible approach. Do you reckon you could take the lead implementing this factorisation?
